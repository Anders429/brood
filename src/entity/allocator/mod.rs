#[cfg(feature = "serde")]
#[cfg_attr(doc_cfg, doc(cfg(feature = "serde")))]
mod impl_serde;
mod location;
mod locations;
mod slot;

#[cfg(feature = "serde")]
pub(crate) use impl_serde::DeserializeAllocator;
pub(crate) use location::Location;
pub(crate) use locations::Locations;
pub(crate) use slot::Slot;

use crate::{
    archetype,
    entity,
    registry::Registry,
};
use alloc::{
    collections::VecDeque,
    vec::Vec,
};
use by_address::ByThinAddress;
use core::{
    fmt,
    fmt::Debug,
};
use fnv::FnvBuildHasher;
use hashbrown::HashMap;

pub struct Allocator<R>
where
    R: Registry,
{
    pub(crate) slots: Vec<Slot<R>>,
    pub(crate) free: VecDeque<usize>,
}

impl<R> Allocator<R>
where
    R: Registry,
{
    pub(crate) fn new() -> Self {
        Self {
            slots: Vec::new(),
            free: VecDeque::new(),
        }
    }

    pub(crate) fn allocate(&mut self, location: Location<R>) -> entity::Identifier {
        let (index, generation) = if let Some(index) = self.free.pop_front() {
            let slot =
                // SAFETY: `self.free` is guaranteed to contain valid indices within the bounds of
                // `self.slots`.
                unsafe {self.slots.get_unchecked_mut(index)};
            // SAFETY: `self.free` is guaranteed to contain indices for slots that are not active.
            unsafe { slot.activate_unchecked(location) };
            (index, slot.generation)
        } else {
            let index = self.slots.len();
            self.slots.push(Slot::new(location));
            // Generation is always 0 for a new slot.
            (index, 0)
        };

        entity::Identifier::new(index, generation)
    }

    #[inline]
    pub(crate) fn allocate_batch(
        &mut self,
        mut locations: Locations<R>,
    ) -> Vec<entity::Identifier> {
        let mut identifiers = Vec::with_capacity(locations.len());

        // First activate slots that are already allocated.
        while let Some(index) = self.free.pop_front() {
            if locations.is_empty() {
                break;
            }
            let slot =
                // SAFETY: indices within `self.free` are guaranteed to be within the bounds of
                // `self.slots`.
                unsafe { self.slots.get_unchecked_mut(index) };
            // SAFETY: `self.free` is guaranteed to contain indices for slots that are not active.
            // Also, `locations` is already checked above to be nonempty.
            unsafe { slot.activate_unchecked(locations.next().unwrap_unchecked()) };
            identifiers.push(entity::Identifier::new(index, slot.generation));
        }

        // Now allocate the remaining slots.
        let remaining_locations = locations.len();
        let slots_len = self.slots.len();
        self.slots
            .extend(locations.map(|location| Slot::new(location)));
        identifiers.extend(
            (0..remaining_locations).map(|index| entity::Identifier::new(slots_len + index, 0)),
        );

        identifiers
    }

    pub(crate) fn get(&self, identifier: entity::Identifier) -> Option<Location<R>> {
        let slot = self.slots.get(identifier.index)?;
        if slot.generation == identifier.generation {
            slot.location
        } else {
            None
        }
    }

    pub(crate) fn is_active(&self, identifier: entity::Identifier) -> bool {
        if let Some(slot) = self.slots.get(identifier.index) {
            if slot.is_active() && slot.generation == identifier.generation {
                return true;
            }
        }
        false
    }

    /// Free the entity allocation identified by `identifier`, skipping checks for whether the
    /// allocation exists.
    ///
    /// # Safety
    /// `identifier` must be for a valid, currently allocated entity.
    pub(crate) unsafe fn free_unchecked(&mut self, identifier: entity::Identifier) {
        let slot =
            // SAFETY: `identifier` is guaranteed by the safety contract of this method to identify
            // a valid entity. Therefore, its `index` will correspond to a valid value within
            // `self.slots`.
            unsafe {self.slots.get_unchecked_mut(identifier.index)};
        slot.deactivate();
        self.free.push_back(identifier.index);
    }

    /// Update the location of the entity identified by `identifier`, skipping checks for whether
    /// the allocation exists.
    ///
    /// # Safety
    /// `identifier` must be for a valid, currently allocated entity.
    pub(crate) unsafe fn modify_location_unchecked(
        &mut self,
        identifier: entity::Identifier,
        location: Location<R>,
    ) {
        // SAFETY: `identifier` is guaranteed by the safety contract of this method to identify a
        // valid entity. Therefore, its `index` will correspond to a valid value within
        // `self.slots`.
        unsafe { self.slots.get_unchecked_mut(identifier.index) }.location = Some(location);
    }

    /// Update the location's index of the entity identified by `identifier`, skipping checks for
    /// whether the allocation exists.
    ///
    /// This should be used when an entity's location within an archetype table has changed.
    /// Calling this method ensures the allocator's mapping of where entities currently are stays
    /// up to date.
    ///
    /// # Safety
    /// `identifier` must be for a valid, currently allocated entity.
    pub(crate) unsafe fn modify_location_index_unchecked(
        &mut self,
        identifier: entity::Identifier,
        index: usize,
    ) {
        // SAFETY: `identifier` is guaranteed by the safety contract of this method to identify a
        // valid active entity. Therefore, its `index` will correspond to a valid active value
        // within `self.slots`.
        unsafe {
            self.slots
                .get_unchecked_mut(identifier.index)
                .location
                .as_mut()
                .unwrap_unchecked()
        }
        .index = index;
    }

    /// Decrease the allocated capacity to the smallest amount required for the stored data.
    ///
    /// This may not decrease to the most optimal value, as the shrinking is dependent on the
    /// allocator.
    ///
    /// Note that this only affects the list of currently free indexes. Slots are never removed, so
    /// there is no need to shrink them.
    pub(crate) fn shrink_to_fit(&mut self) {
        self.free.shrink_to_fit();
    }

    /// Clone the entity allocator, using `identifier_map` to replace old archetype identifiers
    /// with new ones.
    ///
    /// # Safety
    /// `identifier_map` must contain entries for every archetype referenced in this entity
    /// allocator.
    pub(crate) unsafe fn clone(
        &self,
        identifier_map: &HashMap<ByThinAddress<&[u8]>, archetype::IdentifierRef<R>, FnvBuildHasher>,
    ) -> Self {
        Self {
            slots: self
                .slots
                .iter()
                .map(|slot|
                // SAFETY: `identifier_map` contains an entry for the archetype referenced in
                // `slot`, if there is one.
                unsafe {slot.clone_with_new_identifier(identifier_map)})
                .collect(),
            free: self.free.clone(),
        }
    }

    /// Clone from another entity allocator into this one, using `identifier_map` to replace old
    /// archetype identifiers with new ones.
    ///
    /// This reuses the existing allocations.
    ///
    /// # Safety
    /// `identifier_map` must contain entries for every archetype referenced in this entity
    /// allocator.
    pub(crate) unsafe fn clone_from(
        &mut self,
        source: &Self,
        identifier_map: &HashMap<ByThinAddress<&[u8]>, archetype::IdentifierRef<R>, FnvBuildHasher>,
    ) {
        self.slots.clear();
        self.slots.extend(source.slots.iter().map(|slot|
            // SAFETY: `identifier_map` contains an entry for the archetype referenced in
            // `slot`, if there is one.
            unsafe {slot.clone_with_new_identifier(identifier_map)}));

        self.free.clone_from(&source.free);
    }
}

impl<R> Debug for Allocator<R>
where
    R: Registry,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Allocator")
            .field("slots", &self.slots)
            .field("free", &self.free)
            .finish()
    }
}

impl<R> PartialEq for Allocator<R>
where
    R: Registry,
{
    fn eq(&self, other: &Self) -> bool {
        self.slots == other.slots && self.free == other.free
    }
}
