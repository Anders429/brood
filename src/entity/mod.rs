//! A heterogeneous list of [`Component`]s stored within a [`World`].
//!
//! [`Entity`]s are most often defined using the [`entity!`] macro. The items contained within this
//! module should rarely be needed in user code, apart from [`Identifier`].
//!
//! `Entity`s are stored within [`World`]s to allow efficient querying and iteration with other
//! entities of similar components. Since entities are defined as heterogeneous lists, they can be
//! made of an arbitrary number of components. `World`s can store entities made up of any
//! combination of components, so long as those components are stored in the `World`'s
//! [`Registry`].
//!
//! # Example
//! ``` rust
//! use brood::{entity, registry, World};
//!
//! // Define components.
//! struct Foo(usize);
//! struct Bar(bool);
//!
//! type Registry = registry!(Foo, Bar);
//!
//! let mut world = World::<Registry>::new();
//!
//! // Store an entity containing both `Foo` and `Bar`.
//! world.insert(entity!(Foo(42), Bar(false)));
//!
//! // Store an entity containing only `Bar`.
//! world.insert(entity!(Bar(true)));
//!
//! // Store an entity containing zero components.
//! world.insert(entity!());
//! ```
//!
//! [`Entity`]: crate::entity::Entity
//! [`entity!`]: crate::entity!
//! [`Identifier`]: crate::entity::Identifier
//! [`Registry`]: crate::registry::Registry
//! [`World`]: crate::world::World

pub(crate) mod allocator;

mod identifier;
mod seal;

pub use identifier::Identifier;

pub(crate) use allocator::Allocator;

use crate::{component::Component, hlist::define_null};
use seal::Seal;

define_null!();

/// A heterogeneous list of [`Component`]s.
///
/// Entities are stored within [`World`]s. In order for an entity to be able to be stored within a
/// `World`, that `World`'s [`Registry`] must include the `Component`s that make up an entity.
///
/// Note that entities must consist of unique component types. Duplicate components are not
/// supported. When multiple components of the same type are included in an entity, a `World` will
/// only store one of those components.
///
/// # Example
/// ``` rust
/// use brood::entity;
///
/// // Define components.
/// struct Foo(usize);
/// struct Bar(bool);
///
/// let entity = entity!(Foo(42), Bar(true));
/// ```
///
/// [`Component`]: crate::component::Component
/// [`Registry`]: crate::registry::Registry
/// [`World`]: crate::world::World
pub trait Entity: Seal {}

impl Entity for Null {}

impl<C, E> Entity for (C, E)
where
    C: Component,
    E: Entity,
{
}

/// Creates an entity from the provided components.
///
/// This macro allows an enity to be defined without needing to manually create a heterogeneous
/// list of components. Given an arbitrary number of components, this macro will arrange them into
/// nested tuples forming a heterogeneous list.
///
/// # Example
/// ``` rust
/// use brood::entity;
///
/// // Define components `Foo` and `Bar`.
/// struct Foo(u16);
/// struct Bar(f32);
///
/// // Define an entity containing an instance of `Foo` and `Bar`.
/// let my_entity = entity!(Foo(42), Bar(1.5));
/// ```
#[macro_export]
macro_rules! entity {
    ($component:expr $(,$components:expr)* $(,)?) => {
        ($component, $crate::entity!($($components,)*))
    };
    () => {
        $crate::entity::Null
    };
}
