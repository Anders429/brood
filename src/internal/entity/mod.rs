mod byte_length;
mod length;
mod storage;

use crate::{component::Component, entity::{NullEntities, NullEntity}};
use alloc::vec::Vec;
use byte_length::EntityByteLength;
use length::{EntitiesLength, EntityLength};
use storage::{EntitiesStorage, EntityStorage};

pub trait EntitySeal: EntityByteLength + EntityLength + EntityStorage {}

impl EntitySeal for NullEntity {}

impl<C, E> EntitySeal for (C, E)
where
    C: Component,
    E: EntitySeal,
{
}

pub trait EntitiesSeal: EntitiesLength + EntitiesStorage {}

impl EntitiesSeal for NullEntities {}

impl<C, E> EntitiesSeal for (Vec<C>, E) where C: Component, E: EntitiesSeal, {}
