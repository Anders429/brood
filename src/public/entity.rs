use crate::{component::Component, internal::entity::{EntitiesSeal, EntitySeal}};
use core::any::Any;
use alloc::vec::Vec;

pub struct NullEntity;

pub trait Entity: EntitySeal + Any {}

impl Entity for NullEntity {}

impl<C, E> Entity for (C, E)
where
    C: Component,
    E: Entity,
{
}

#[macro_export]
macro_rules! entity {
    ($component:expr $(,$components:expr)* $(,)?) => {
        ($component, entity!($($components,)*))
    };
    () => {
        $crate::entity::NullEntity
    };
}

pub struct NullEntities;

pub trait Entities: EntitiesSeal {}

impl Entities for NullEntities {}

impl<C, E> Entities for (Vec<C>, E) where C: Component, E: Entities {}

#[macro_export]
macro_rules! entities {
    (($component:expr $(,$components:expr)* $(,)?); $n:expr) => {
        ($crate::reexports::vec![$component; $n], entities!(($($components),*); $n))
    };
    ($($($components:expr),*),*) => {};
    ((); $n:expr) => {
        $crate::entity::NullEntities
    };
    () => {
        $crate::entity::NullEntities
    };
}
