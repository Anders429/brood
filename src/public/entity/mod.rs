mod identifier;

pub use identifier::EntityIdentifier;

use crate::{component::Component, internal::entity::EntitySeal};
use core::any::Any;

pub struct NullEntity;

pub trait Entity: EntitySeal + Any {}

impl Entity for NullEntity {}

impl<C, E> Entity for (C, E)
where
    C: Component,
    E: Entity,
{
}

#[macro_export]
macro_rules! entity {
    ($component:expr $(,$components:expr)* $(,)?) => {
        ($component, entity!($($components,)*))
    };
    () => {
        $crate::entity::NullEntity
    };
}
