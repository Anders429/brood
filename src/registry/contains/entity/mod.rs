//! Provides a `ContainsEntity` trait to indicate that a registry contains an entity.
//!
//! This containment relationship directly implies that a canonical version of an entity, with
//! respect to the registry, can be derived. That canonical form is obtainable through this trait.

mod sealed;

pub(crate) use sealed::Sealed;

/// Indicates that all of an entity's components are contained in the registry.
///
/// This allows reordering the components of the entity into the canonical ordering defined by the
/// registry.
///
/// If the entity contains components not in this registry, attempting to use this trait will
/// result in a compiler error, since the trait won't be implemented for the combination of entity
/// and registry.
pub trait ContainsEntity<Entity, Indices>: Sealed<Entity, Indices> {}

impl<Registry, Entity, Indices> ContainsEntity<Entity, Indices> for Registry where
    Registry: Sealed<Entity, Indices>
{
}
