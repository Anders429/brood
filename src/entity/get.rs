//! Extracts a component from an entity.
//!
//! This is defined for every entity that contains a given component. If an entity does not contain
//! a component, calling `get()` for that component will fail to compile.

/// Type marker for the location of a component.
///
/// This does not actually have to be specified when calling `get()`. The compiler can infer its
/// location.
pub enum Index {}

/// Extracts a component from an entity.
///
/// Removes the requested component, returning the remainder of the entity with the component
/// removed.
///
/// If a component is not contained in the entity, attempting to `get()` it with this trait will
/// fail to compile. This is because it is defined recursively on either the component itself or
/// lists that have the component within them.
///
/// This is generic over a component `T` and an index `I` (denoting the location of the component).
pub trait Get<T, I> {
    /// The entity returned after the component `T` is extracted.
    type Remainder;

    /// Remove the component `T` from the entity.
    ///
    /// Consumes the entity, returning the component and the entity with the component removed.
    fn get(self) -> (T, Self::Remainder);
}

impl<E, T> Get<T, Index> for (T, E) {
    type Remainder = E;

    fn get(self) -> (T, Self::Remainder) {
        self
    }
}

impl<C, E, I, T> Get<T, (I,)> for (C, E)
where
    E: Get<T, I>,
{
    type Remainder = (C, <E as Get<T, I>>::Remainder);

    fn get(self) -> (T, Self::Remainder) {
        let (target, remainder) = self.1.get();
        (target, (self.0, remainder))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::entity;

    #[derive(Debug, Eq, PartialEq)]
    struct A;
    #[derive(Debug, Eq, PartialEq)]
    struct B;
    #[derive(Debug, Eq, PartialEq)]
    struct C;

    #[test]
    fn get_a() {
        assert_eq!(entity!(A, B, C).get(), (A, entity!(B, C)));
    }

    #[test]
    fn get_b() {
        assert_eq!(entity!(A, B, C).get(), (B, entity!(A, C)));
    }

    #[test]
    fn get_c() {
        assert_eq!(entity!(A, B, C).get(), (C, entity!(A, B)));
    }
}
