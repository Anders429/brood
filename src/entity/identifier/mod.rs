#[cfg(feature = "serde")]
#[cfg_attr(doc_cfg, doc(cfg(feature = "serde")))]
mod impl_serde;

/// A unique identifier for an entity.
///
/// An `Identifier` can be used to reference an entity that is stored within a [`World`].
/// `Identifier`s are normally obtained when inserting an entity into a `World`, although they can
/// also be obtained through a [`query`] on a `World` by providing `Identifier` as a [`View`].
///
/// # Example
/// ``` rust
/// use brood::{entity, registry, World};
///
/// // Define components.
/// struct Foo(usize);
/// struct Bar(bool);
///
/// type Registry = registry!(Foo, Bar);
///
/// let mut world = World::<Registry>::new();
///
/// // An identifier is returned on insertion.
/// let entity_identifier = world.push(entity!(Foo(42), Bar(false)));
/// ```
///
/// [`query`]: crate::world::World::query()
/// [`View`]: crate::query::view::View
/// [`World`]: crate::world::World
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub struct Identifier {
    pub(crate) index: usize,
    pub(crate) generation: u64,
}

impl Identifier {
    pub(crate) fn new(index: usize, generation: u64) -> Self {
        Self { index, generation }
    }
}

#[cfg(test)]
mod tests {
    use crate::entity::Identifier;

    #[test]
    fn new_index() {
        let identifier = Identifier::new(1, 2);

        assert_eq!(identifier.index, 1);
    }

    #[test]
    fn new_generation() {
        let identifier = Identifier::new(1, 2);

        assert_eq!(identifier.generation, 2);
    }
}
