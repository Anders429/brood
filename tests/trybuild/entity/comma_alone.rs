use brood::entity;

fn main() {
    let entity = entity!(,);
}
