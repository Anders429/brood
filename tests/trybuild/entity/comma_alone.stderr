error: no rules expected the token `,`
 --> tests/trybuild/entity/comma_alone.rs:4:26
  |
4 |     let entity = entity!(,);
  |                          ^ no rules expected this token in macro call
  |
note: while trying to match meta-variable `$component:expr`
 --> src/entity/mod.rs
  |
  |     ($component:expr $(,$components:expr)* $(,)?) => {
  |      ^^^^^^^^^^^^^^^
